In [7]:
# Created by @eagskunst 2020
import pandas as pd
import numpy as np

In [8]:
gender_df = pd.read_csv('db/gender_submission.csv')
gender_df

PassengerId  Survived
0            892         0
1            893         1
2            894         0
3            895         0
4            896         1
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [9]:
people_df = pd.read_csv('db/test.csv')
people_df

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [10]:
train_df = pd.read_csv('db/train.csv')
train_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

## 1. How many people were on the Titanic?

In [11]:
df_final = pd.merge(gender_df, people_df, how='inner', on='PassengerId')
df_final = pd.concat([df_final, train_df]).reset_index(drop=True).sort_index() #removes duplicate index and sort is
df_final[df_final['Name'].isna()] #check if there are null names

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [12]:
df_final['Name'].value_counts() #check if there are dup names

Connolly, Miss. Kate                                2
Kelly, Mr. James                                    2
Barry, Miss. Julia                                  1
Abrahamsson, Mr. Abraham August Johannes            1
Peacock, Master. Alfred Edward                      1
                                                   ..
Mayne, Mlle. Berthe Antonine ("Mrs de Villiers")    1
Wright, Miss. Marion                                1
Mallet, Mr. Albert                                  1
Elias, Mr. Joseph Jr                                1
Simonius-Blumer, Col. Oberst Alfons                 1
Name: Name, Length: 1307, dtype: int64

In [13]:
#df_final['Name'] = df_final['Name'].drop_duplicates()  #drop dup names
df_final['Name'].value_counts()

Connolly, Miss. Kate                                2
Kelly, Mr. James                                    2
Barry, Miss. Julia                                  1
Abrahamsson, Mr. Abraham August Johannes            1
Peacock, Master. Alfred Edward                      1
                                                   ..
Mayne, Mlle. Berthe Antonine ("Mrs de Villiers")    1
Wright, Miss. Marion                                1
Mallet, Mr. Albert                                  1
Elias, Mr. Joseph Jr                                1
Simonius-Blumer, Col. Oberst Alfons                 1
Name: Name, Length: 1307, dtype: int64

In [14]:
df_final['Name'].count()

1309

### There were 1309 persons at the Titanic

## 2. How many man and woman survived?

In [15]:
df_surv = df_final.copy(deep=True)[['Sex', 'Survived']]
df_surv['Sex'] = pd.Categorical(df_final['Sex'])
df_surv.groupby(['Sex']).count()

Survived
Sex             
female       466
male         843

In [16]:
surv_mask = df_surv['Survived'] == 1
df_surv = df_surv[surv_mask]

In [17]:
df_surv.groupby(['Sex']).count()

Survived
Sex             
female       385
male         109


### 358 females and 109 males survived

## 3. Top 10 age that survived and top 10 that did not

In [18]:
df_ages = df_final.copy(deep=True)[['Age', 'Survived']]
df_ages

Age  Survived
0     34.5         0
1     47.0         1
2     62.0         0
3     27.0         0
4     22.0         1
...    ...       ...
1304  27.0         0
1305  19.0         1
1306   NaN         0
1307  26.0         1
1308  32.0         0

[1309 rows x 2 columns]

In [19]:
df_ages = df_ages.dropna()

In [20]:
survived_mask = df_ages['Survived'] == 1
df_ages_surv = df_ages[survived_mask]
grouped_df = df_ages_surv.groupby(['Age'])
grouped_df.sum()

Survived
Age            
0.17          1
0.42          1
0.67          1
0.75          2
0.83          2
...         ...
62.00         2
63.00         3
64.00         2
76.00         1
80.00         1

[71 rows x 1 columns]

In [21]:
grouped_df = grouped_df.sum()
df_ages_surv = grouped_df.sort_values('Survived', ascending=False)
df_ages_surv

Survived
Age            
22.00        21
24.00        20
30.00        16
36.00        16
18.00        16
...         ...
11.00         1
10.00         1
7.00          1
0.67          1
80.00         1

[71 rows x 1 columns]

In [22]:
df_ages_surv.head(10)

Survived
Age           
22.0        21
24.0        20
30.0        16
36.0        16
18.0        16
27.0        15
35.0        14
29.0        13
19.0        12
45.0        12

In [23]:
not_surv_mask = df_ages['Survived'] == 0
df_ages_not_surv = df_ages[not_surv_mask].copy(deep=True)
df_ages_not_surv['Survived'] = df_ages_not_surv['Survived'].map(lambda n: 1)
df_ages_not_surv.rename(columns={'Survived': 'not survive'}, inplace=True)
df_ages_not_surv = (
    df_ages_not_surv.groupby(['Age'])
    .sum()
    .sort_values('not survive', ascending=False)
)
df_ages_not_surv.head(10)

not survive
Age              
21.0           33
25.0           27
24.0           27
30.0           24
28.0           23
18.0           23
22.0           22
26.0           20
19.0           17
29.0           17

_Format: age (quantity)_

### Top 10 ages that survived:

1. 22 (21) 
2. 24 (20) 
3. 30 (16) 
4. 36 (16) 
5. 18 (16) 
6. 27 (15) 
7. 35 (14)
8. 29 (13)
9. 19 (12)
10. 45 (12)

### Top 10 ages that did not survived:

1. 21 (33) 
2. 25 (27) 
3. 24 (27) 
4. 30 (24) 
5. 28 (23)
6. 18 (23) 
7. 22 (22) 
8. 26 (20)
9. 19 (17)
10. 29 (17)


## 4. How many titles/charges where in the ship?

In [24]:
df_final['Name']

0                                   Kelly, Mr. James
1                   Wilkes, Mrs. James (Ellen Needs)
2                          Myles, Mr. Thomas Francis
3                                   Wirz, Mr. Albert
4       Hirvonen, Mrs. Alexander (Helga E Lindqvist)
                            ...                     
1304                           Montvila, Rev. Juozas
1305                    Graham, Miss. Margaret Edith
1306        Johnston, Miss. Catherine Helen "Carrie"
1307                           Behr, Mr. Karl Howell
1308                             Dooley, Mr. Patrick
Name: Name, Length: 1309, dtype: object

In [25]:
person_charges = df_final['Name'].str.findall('(?<=, ).+')
person_charges = person_charges.apply(lambda collection: ''.join(collection))
person_charges

0                                Mr. James
1                 Mrs. James (Ellen Needs)
2                       Mr. Thomas Francis
3                               Mr. Albert
4       Mrs. Alexander (Helga E Lindqvist)
                       ...                
1304                           Rev. Juozas
1305                  Miss. Margaret Edith
1306        Miss. Catherine Helen "Carrie"
1307                       Mr. Karl Howell
1308                           Mr. Patrick
Name: Name, Length: 1309, dtype: object

In [26]:
person_charges = (person_charges
                  .str
                  .split('.')
                  .apply(lambda words: ''.join(words[0]))                  
)
person_charges

0         Mr
1        Mrs
2         Mr
3         Mr
4        Mrs
        ... 
1304     Rev
1305    Miss
1306    Miss
1307      Mr
1308      Mr
Name: Name, Length: 1309, dtype: object

In [27]:
print(person_charges.unique())
print(len(person_charges.unique()))

['Mr' 'Mrs' 'Miss' 'Master' 'Ms' 'Col' 'Rev' 'Dr' 'Dona' 'Don' 'Mme'
 'Major' 'Lady' 'Sir' 'Mlle' 'Capt' 'the Countess' 'Jonkheer']
18


### There are 18 titles/charges between the persons that were on the ship:
1. Mr
2. Mrs
3. Miss
4. Master
5. Ms
6. Col
7. Rev
8. Dr
9. Dona
10. Don
11. Mme
12. Major
13. Lady
14. Sir
15. Mlle
16. Capt
17. The Countess
18. Jonkheer

## 5. Total of the tickets prices in USD

In [28]:
df_final['Fare'].sum() #Fare = tarifa

43550.4869

### Total price of the tickets that were bought is 43550.4869 USD